plugins/auth-backend-module-oidc-provider/src/authenticator.test.ts

```diff
@@ -0,0 +1,441 @@
+  });
+
+  describe('#authenticate', () => {
+    let handlerRequest: OAuthAuthenticatorAuthenticateInput;
+
+    beforeEach(() => {
+      handlerRequest = {
+        req: {
+          method: 'GET',
+          url: `https://test?code=authorization_code&state=${encodeOAuthState(
+            oauthState,
+          )}`,
+          session: {
+            'oidc:oidc.test': {
+              state: encodeOAuthState(oauthState),
+            },
+          },
+        } as unknown as express.Request,
+      };
+    });
+
+    it('exchanges authorization code for access token', async () => {
+      const handlerResponse = await oidcAuthenticator.authenticate(
+        handlerRequest,
+        implementation,
+      );
+      const accessToken = handlerResponse.session.accessToken;
+
+      expect(accessToken).toEqual('accessToken');
+    });
+
+    it('exchanges authorization code for refresh token', async () => {
+      const handlerResponse = await oidcAuthenticator.authenticate(
+        handlerRequest,
+        implementation,
+      );
+      const refreshToken = handlerResponse.session.refreshToken;
+
+      expect(refreshToken).toEqual('refreshToken');
+    });
+
+    it('returns granted scope', async () => {
+      const handlerResponse = await oidcAuthenticator.authenticate(
+        handlerRequest,
+        implementation,
+      );
+      const responseScope = handlerResponse.session.scope;
+
+      expect(responseScope).toEqual('testScope');
+    });
+
+    it('returns a default session.tokentype field', async () => {
+      const handlerResponse = await oidcAuthenticator.authenticate(
+        handlerRequest,
+        implementation,
+      );
+      const tokenType = handlerResponse.session.tokenType;
+
+      expect(tokenType).toEqual('bearer');
+    });
+
+    it('returns defined fullProfile with picture and email', async () => {
+      const handlerResponse = await oidcAuthenticator.authenticate(
+        handlerRequest,
+        implementation,
+      );
+      const displayName = handlerResponse.fullProfile.displayName;
+      const email = handlerResponse.fullProfile.emails![0].value;
+      const picture = handlerResponse.fullProfile.photos![0].value;
+
+      expect(displayName).toEqual('Alice Adams');
+      expect(email).toEqual('alice@test.com');
+      expect(picture).toEqual('http://testPictureUrl/photo.jpg');
+    });
+
+    it('returns defined response with an idToken', async () => {
+      const handlerResponse = await oidcAuthenticator.authenticate(
+        handlerRequest,
+        implementation,
+      );
+
+      expect(handlerResponse).toMatchObject({
+        fullProfile: {
+          displayName: 'Alice Adams',
+          id: 'test',
+          provider: 'oidc',
+        },
+        session: {
+          accessToken: 'accessToken',
+          expiresInSeconds: 3600,
+          idToken,
+          refreshToken: 'refreshToken',
+          scope: 'testScope',
+          tokenType: 'bearer',
+        },
+      });
```

Comment:
...a second may have passed on the clock by the time you get here, I think. Maybe better to use a matcher with some wiggle room, something like
```suggestion
      expect(handlerResponse).toMatchObject({
        fullProfile: {
          displayName: 'Alice Adams',
          id: 'test',
          provider: 'oidc',
        },
        session: {
          accessToken: 'accessToken',
          expiresInSeconds: expect.any(Number),
          idToken,
          refreshToken: 'refreshToken',
          scope: 'testScope',
          tokenType: 'bearer',
        },
      });
      expect(math.abs(handlerResponse.session.expiresInSeconds - 3600)).toBeLessThan(5);
```

Reply:
Added change. Thanks. Done.